portfolio/src/main/webapp/script.js

```diff
@@ -29,100 +28,201 @@ function addRandomGreeting() {
+    return indexArr;
+}
+
+/**
+ * Clears the current highlighted text nodes
+ * Will optimize to actually remove the span rather than the class
+ */
+function clear(){
+
+    var highlightedList = document.querySelectorAll('.highlight');
 
+    highlightedList.forEach(function(el) {
+        el.classList.remove('highlight');
+    });
+    
 }
 
-function retrieveId(string, source){
+/**
+ * Handles clearing old nodes and end of search run,
+ * as well as calls the highlighting and scrolling fns
+ */
+function treeWalkerHelper(matches, input, index){
 
-    var searchForId = string.search("id=\"");
-    if (!searchForId){
+    if (index === matches.length && index != 0){
+        clear();
+        setHidden('nextbutton', true);
+        document.body.scrollIntoView({behavior: "smooth"}); //back to the top
         return;
     }
-    const startIndex = searchForId+4;
-    var retString = string.charAt(startIndex);
 
-    for(var i = startIndex+1; string.charAt(i) != "\""; i++){
-        retString += string.charAt(i);
+    if (index > 0){
+        clear();
     }
 
-    return retString;
+    console.log(matches[index]);
+    highlighter(matches[index], input);
+
+    var els = document.getElementsByClassName('highlight');
+    els[0].scrollIntoView({behavior: "smooth"}); 
 }
 
-function runSearch(input){
+/**
+ * Highlights a given text node
+ */
+function highlighter(node, input){
 
-    var source = retrieveHtml();
+    try {
+        var foundIndices = getIndicesOf(input, node.textContent);
+    } catch(err) {
+        console.log("indices" + foundIndices);
+    }
 
-    var foundIndex = source.indexOf(input);
-    if (foundIndex == -1){return};
+    var stringBefore = node.textContent.substring(0, foundIndices[0]);
+    var stringAfter = node.textContent.substring(input.length+foundIndices[0]);
+    var highlightedString = node.textContent.substring(foundIndices[0], input.length+foundIndices[0]);
+    
+    var newNode = document.createElement('span');
+    newNode.setAttribute('class', 'highlight');
+    newNode.textContent = highlightedString;
 
-    var closestTag = getClosestTag(foundIndex, source); 
-    var gottenId = retrieveId(closestTag, source);
+    node.replaceWith(stringBefore, newNode, stringAfter);
+}
 
+/**
+ * Walks the tree to find matches, displays the next button,
+ * runs the search for each match
+ */
+function runSearch(input){
+    
+    var matchingNodes = treeWalker(document.body, input);
+
+    //show the next button
+    setHidden('nextbutton', false);
 
-    if(gottenId == null){
-        alert("gottenId == null");
+    if (matchingNodes.length == 0){
+        setHidden('nextbutton', true);
+        shakeSearchBar();
         return;
     }
-    document.getElementById(gottenId).scrollIntoView({behavior: "smooth"}); // no support for safari or ie
-    
-
-}
 
+    var counter = 0;
+    //Enter searching mode
+    var el = document.getElementById("nextbutton");
+    el.addEventListener("click", function(){ 
```

Comment:
You never remove this event listener. After you type the search twice, clicking next will call this method twice with two different contexts (the one from your first search and the one from your second search). You should probably make file level variables for your counter and input. Then you can reset them whenever you search but only have to assign this event listener once.

Reply:
Thank you so much for pointing this out. I think I'm also going to read up more on event listener.